### PR TITLE
fix: let oss support archive none when oss-key is a directory

### PR DESCRIPTION
**Pull request by jibuji**
I used this template to test it and see the result on oss:

```yaml
  - name: oss-getter-and-uploader
    inputs:
      artifacts:
      - name: oss-data
        path: "/mnt/input"
        oss:
          key: "some/oss/dir"
        archive:
          none: {}
    container:
      image: 'alpine:latest'
      command: ['/bin/sh', '-c']
      args: ["cp -rf /mnt/input /mnt/output"]
      volumeMounts:
        - name: out
          mountPath: /mnt
    volumes:
      - name: out
        emptyDir: { }
    outputs:
      artifacts:
      - name: output
        path: "/mnt/output"
        oss:
          key: "some/oss/dir2"
        archive:
          none: {}
```

### CODE REVIEW

**Comment by terrytangyuan** [comment]
Thanks for this. Could you also test non-directory cases?

**Comment by terrytangyuan** [comment]
Fixes https://github.com/argoproj/argo-workflows/issues/6293

**Comment by jibuji** [comment]
I modified the code to support save empty dir to oss. And also, I have tested more cases, including file, empty directory, none empty directory, etc. The full test yaml is here:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: test-archive-none
spec:
  entrypoint: oss-getter-and-uploader
  templates:
  - name: oss-getter-and-uploader
    inputs:
      artifacts:
      - name: oss-none-empty-dir
        path: "/mnt/input/none-empty"
        oss:
          key: "some/none-empty"
        archive:
          none: {}
      - name: oss-empty-dir
        path: "/mnt/input/empty-dir"
        oss:
          key: "some/emptydir"
        archive:
          none: {}
      - name: oss-file
        path: "/mnt/input/file"
        oss:
          key: "some/file"
        archive:
          none: {}
      - name: oss-3level
        path: "/mnt/input/3level"
        oss:
          key: "some/3level-dir"
        archive:
          none: {}
    container:
      image: 'alpine:latest'
      command: ['/bin/sh', '-c']
      args: ["cp -rf /mnt/input /mnt/output"]
      volumeMounts:
        - name: out
          mountPath: /mnt
    volumes:
      - name: out
        emptyDir: { }
    outputs:
      artifacts:
      - name: oss-none-empty-dir
        path: "/mnt/output/none-empty"
        oss:
          key: "some2/none-empty"
        archive:
          none: {}
      - name: oss-empty-dir
        path: "/mnt/output/empty-dir"
        oss:
          key: "some2/emptydir"
        archive:
          none: {}
      - name: oss-file
        path: "/mnt/output/file"
        oss:
          key: "some2/file"
        archive:
          none: {}
      - name: oss-3level
        path: "/mnt/output/3level"
        oss:
          key: "some2/3level-dir"
        archive:
          none: {}

```

File: workflow/artifacts/oss/oss.go [hunk 1]
```diff
@@ -59,21 +62,25 @@ func (ossDriver *ArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string)
...
-			err = bucket.GetObjectToFile(objectName+"/", path)
+			// If we get here, the error was a NoSuchKey. The key might be a oss "directory"
+			isDir, err := IsOssDirectory(bucket, objectName)
 			if err != nil {
-				return !isTransientOSSErr(err), err
+				return !isTransientOSSErr(err), fmt.Errorf("failed to test if %s/%s is a directory: %w", bucketName, objectName, err)
+			}
+			if !isDir {
+				// It's neither a file, nor a directory. Return the original NoSuchKey error
+				return false, errors.New(errors.CodeNotFound, err.Error())
```
Comment (terrytangyuan) [inline workflow/artifacts/oss/oss.go:79]: This `err` is overwritten as part of the `IsOssDirectory` check though so the original error is masked out

File: workflow/artifacts/oss/oss.go [hunk 2]
```diff
@@ -212,3 +223,136 @@ func isTransientOSSErr(err error) bool {
...
+	if err != nil {
+		return false, err
+	}
+	if len(rst.CommonPrefixes)+len(rst.Objects) > 0 {
+		return true, nil
+	}
+	return false, nil
+}
+
+// GetOssDirectory download a oss "directory" to local path
```
Comment (terrytangyuan) [inline workflow/artifacts/oss/oss.go:296]:
```suggestion
// GetOssDirectory download an OSS "directory" to local path
```

Comment (terrytangyuan) [reply workflow/artifacts/oss/oss.go:296]: Similar in other places too. Please check.

File: workflow/artifacts/oss/oss.go [hunk 3]
```diff
@@ -212,3 +223,136 @@ func isTransientOSSErr(err error) bool {
...
+// ListOssDirectory lists all the files which are the descendants of the specified objectKey, if a file has suffix '/', then it is an oss directory
+func ListOssDirectory(bucket *oss.Bucket, objectKey string) (files []string, err error) {
+	if objectKey != "" {
+		if !strings.HasSuffix(objectKey, "/") {
+			objectKey += "/"
+		}
+	}
+
+	pre := oss.Prefix(objectKey)
+	const batchCount = 50
```
Comment (terrytangyuan) [inline workflow/artifacts/oss/oss.go:339]: Could you add a comment on why this is hard-coded

File: workflow/artifacts/oss/oss.go (hunk 1 above)
Comment (jibuji) [reply workflow/artifacts/oss/oss.go:79]: You are right, I'll fix it.

File: workflow/artifacts/oss/oss.go (hunk 3 above)
Comment (jibuji) [reply workflow/artifacts/oss/oss.go:339]: this hard-coded "50" has no much meaning, it is just a number sounds reasonable pop up from my head when I was coding it. 50 is not too big that would cause timeout, and not two small that would cause too much network requests.
In fact, if `oss.MaxKeys` is omitted, 100 will be used internally by default.
Maybe I should not explicitly set this option here, I will remove it later.

Comment (terrytangyuan) [reply workflow/artifacts/oss/oss.go:339]: Yes let's remove it and use the default unless really necessary

File: workflow/artifacts/oss/oss.go [hunk 4]
```diff
@@ -212,3 +223,135 @@ func isTransientOSSErr(err error) bool {
...
+			log.Warnf("oss list object(%s) error: %v", objectKey, err)
+			return files, err
+		}
+		for _, obj := range lor.Objects {
+			files = append(files, obj.Key)
+		}
+
+		marker = oss.Marker(lor.NextMarker)
+		if !lor.IsTruncated {
+			//all done
```
Comment (terrytangyuan) [inline workflow/artifacts/oss/oss.go:352]: Remove unnecessary comments

File: workflow/artifacts/oss/oss.go [hunk 5]
```diff
@@ -212,3 +223,135 @@ func isTransientOSSErr(err error) bool {
...
+
+		err = putFile(bucket, fObjectName, fpath)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+}
+
+//IsOssErrCode tests if an err is an oss.ServiceError with the specified code
```
Comment (terrytangyuan) [inline workflow/artifacts/oss/oss.go:267]: Space after //

File: workflow/artifacts/oss/oss.go [hunk 6]
```diff
@@ -212,3 +223,135 @@ func isTransientOSSErr(err error) bool {
...
+			return errors.InternalWrapError(err)
+		}
+		// build the name to be used in OSS
+		nameInDir, err := filepath.Rel(dir, fpath)
+		if err != nil {
+			return errors.InternalWrapError(err)
+		}
+		fObjectName := filepath.Join(objectName, nameInDir)
+		//for local dir, create an OSS dir explicitly;
+		//thus, for an empty local dir, an empty OSS dir is also created.
```
Comment (terrytangyuan) [inline workflow/artifacts/oss/oss.go:244]: Use complete sentences

File: workflow/artifacts/oss/oss.go (hunk 4 above)
Comment (jibuji) [reply workflow/artifacts/oss/oss.go:352]:
```suggestion
```

File: workflow/artifacts/oss/oss.go (hunk 5 above)
Comment (jibuji) [reply workflow/artifacts/oss/oss.go:267]:
```suggestion
// IsOssErrCode tests if an err is an oss.ServiceError with the specified code
```

File: workflow/artifacts/oss/oss.go (hunk 6 above)
Comment (jibuji) [reply workflow/artifacts/oss/oss.go:244]:
```suggestion
		// create an OSS dir explicitly for every local dir, , including empty dirs.
```

**Comment by jibuji** [comment]
I don't know why it is a 'DCO' thing required,  I have done all the  stuff  suggested, please help.

**Comment by terrytangyuan** [comment]
It looks like you included commits from others. Could you please rebase and make sure your branch only contains your changes?

**Comment by jibuji** [comment]
It seems I have messed up the commits. I just followed the DCO suggestions. I need more time to clear out. It is too late tonight, I'll fix it tomorrow.
